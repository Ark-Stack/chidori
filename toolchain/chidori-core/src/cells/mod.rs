pub mod template_cell;
pub mod code_cell;
pub mod web_cell;
pub mod llm_prompt_cell;
pub mod memory_cell;
pub mod embedding_cell;
pub mod code_gen_cell;

use std::cmp::Ordering;
use std::collections::HashMap;
use ts_rs::TS;
use rkyv::{Archive, Deserialize, Serialize};
use serde_json::Value;
use crate::library::std::ai::llm::ChatModelBatch;

#[derive(
    TS,
    Archive,
    serde::Serialize,
    serde::Deserialize,
    Serialize,
    Deserialize,
    Debug,
    PartialEq,
    Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub enum SupportedLanguage {
    PyO3,
    Starlark,
    Deno,
}

#[derive(
    TS,
    Archive,
    serde::Serialize,
    serde::Deserialize,
    Serialize,
    Deserialize,
    Debug,
    PartialEq,
    Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub(crate) struct CodeCell {
    pub(crate) name: Option<String>,
    pub(crate) language: SupportedLanguage,
    pub(crate) source_code: String,
    pub(crate) function_invocation: Option<String>,
}


#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub enum SupportedMemoryProviders {
    InMemory,
}


#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub(crate) struct MemoryCell {
    pub(crate) name: Option<String>,
    pub(crate) provider: SupportedMemoryProviders,
    pub(crate) embedding_function: String,
}


#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub(crate) struct TemplateCell {
    pub(crate) name: Option<String>,
    pub(crate) body: String,
}

#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub(crate) struct WebserviceCellEndpoint {
    pub(crate) method: String,
    pub(crate) route: String,
    pub(crate) depended_function_identity: String,
    pub(crate) arg_mapping: Vec<(String, String)>,
}

#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub(crate) struct WebserviceCell {
    pub(crate) name: Option<String>,
    pub(crate) configuration: String,
    pub(crate) port: u16,
}


#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub(crate) struct ScheduleCell {
    pub(crate) configuration: String,
}


#[derive(
    TS,
    Archive,
    serde::Serialize,
    serde::Deserialize,
    Serialize,
    Deserialize,
    Debug,
    PartialEq,
    Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub enum SupportedModelProviders {
    OpenAI,
}


#[derive(
Default,
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub struct EjectionConfig {
    pub language: String,
    pub mode: String
}




#[derive(
Default,
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub struct LLMPromptCellChatConfiguration {
    pub(crate) import: Option<Vec<String>>,

    #[serde(rename = "fn")]
    pub(crate) function_name: Option<String>,

    pub model: String,
    pub frequency_penalty: Option<f64>,
    pub max_tokens: Option<i64>,
    pub presence_penalty: Option<f64>,
    pub stop: Option<Vec<String>>,
    pub temperature: Option<f64>,
    pub logit_bias: Option<HashMap<String, i32>>,
    pub user: Option<String>,
    pub seed: Option<i64>,
    pub top_p: Option<f64>,
}

#[derive(
    TS,
    Archive,
    serde::Serialize,
    serde::Deserialize,
    Serialize,
    Deserialize,
    Debug,
    PartialEq,
    Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub enum LLMPromptCell {
    Chat {
        function_invocation: bool,
        configuration: LLMPromptCellChatConfiguration,
        name: Option<String>,
        provider: SupportedModelProviders,
        req: String,
    },
    Completion {
        req: String,
    },
}


#[derive(
Default,
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub struct LLMCodeGenCellChatConfiguration {
    #[serde(rename = "fn")]
    pub(crate) function_name: Option<String>,

    pub model: String,
    pub frequency_penalty: Option<f64>,
    pub max_tokens: Option<i64>,
    pub presence_penalty: Option<f64>,
    pub stop: Option<Vec<String>>,
    pub temperature: Option<f64>,
    pub logit_bias: Option<HashMap<String, i32>>,
    pub user: Option<String>,
    pub seed: Option<i64>,
    pub top_p: Option<f64>,

    pub language: Option<String>,
}

#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub struct LLMCodeGenCell {
    pub function_invocation: bool,
    pub configuration: LLMCodeGenCellChatConfiguration,
    pub name: Option<String>,
    pub provider: SupportedModelProviders,
    pub req: String,
}


#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub struct LLMEmbeddingCell {
    pub function_invocation: bool,
    pub configuration: HashMap<String, String>,
    pub name: Option<String>,
    pub req: String,
}

#[derive(
    TS,
    Archive,
    serde::Serialize,
    serde::Deserialize,
    Serialize,
    Deserialize,
    Debug,
    PartialEq,
    Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub enum CellTypes {
    Code(CodeCell),
    CodeGen(LLMCodeGenCell),
    Prompt(LLMPromptCell),
    Embedding(LLMEmbeddingCell),
    Web(WebserviceCell),
    Template(TemplateCell),
    Memory(MemoryCell),
}

impl Eq for CellTypes {

}

impl PartialOrd for CellTypes {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(Ordering::Equal)
    }
}

impl Ord for CellTypes {
    fn cmp(&self, other: &Self) -> Ordering {
        self.partial_cmp(other).unwrap_or(Ordering::Equal)
    }
}

pub fn get_cell_name(cell: &CellTypes) -> &Option<String> {
    match &cell {
        CellTypes::Code(c) => &c.name,
        CellTypes::Prompt(c) => match c {
            LLMPromptCell::Chat { name, .. } => name,
            LLMPromptCell::Completion { .. } => &None,
        },
        CellTypes::Web(c) => &c.name,
        CellTypes::Template(c) => &c.name,
        CellTypes::Memory(c) => &c.name,
        CellTypes::Embedding(c) => &c.name,
        CellTypes::CodeGen(c) => &c.name
    }
}
