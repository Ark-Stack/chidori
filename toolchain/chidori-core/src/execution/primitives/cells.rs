use rkyv::{
    archived_root, check_archived_root,
    ser::{serializers::AllocSerializer, Serializer},
    Archive, Deserialize, Serialize,
};
use serde_json::Value;
use std::collections::HashMap;

#[derive(Archive, Serialize, Deserialize, Debug, PartialEq, Clone)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
pub enum SupportedLanguage {
    Python,
    Starlark,
    Deno,
}

#[derive(Archive, Serialize, Deserialize, Debug, PartialEq, Clone)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
pub(crate) struct CodeCell {
    pub(crate) language: SupportedLanguage,
    pub(crate) source_code: String,
    pub(crate) function_invocation: Option<String>,
}

#[derive(Archive, Serialize, Deserialize, Debug, PartialEq, Clone)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
pub enum SupportedModelProviders {
    OpenAI,
}

#[derive(Archive, Serialize, Deserialize, Debug, PartialEq, Clone)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
pub enum LLMPromptCell {
    Chat {
        path: Option<String>,
        provider: SupportedModelProviders,
        req: String,
    },
    Completion {
        req: String,
    },
    Embedding {
        req: String,
    },
}

#[derive(Archive, Serialize, Deserialize, Debug, PartialEq, Clone)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
pub enum CellTypes {
    Code(CodeCell),
    Prompt(LLMPromptCell),
}
