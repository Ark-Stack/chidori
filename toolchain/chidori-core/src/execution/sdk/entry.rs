/// This is an SDK for building execution graphs. It is designed to be used iteratively.

/// Add a block of code to a runtime execution graph
fn load_module() {}

/// Execute a target directory of code. Specify your entrypoint into the code base.
fn parse_folder() {}
