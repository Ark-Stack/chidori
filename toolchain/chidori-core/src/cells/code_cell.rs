use futures_util::FutureExt;
use crate::cells::{CodeCell, SupportedLanguage};
use crate::execution::execution::ExecutionState;
use crate::execution::primitives::operation::{InputItemConfiguation, InputSignature, InputType, OperationNode, OutputItemConfiguation, OutputSignature};

/// Code cells allow notebooks to evaluate source code in a variety of languages.
#[tracing::instrument]
pub fn code_cell(cell: &CodeCell) -> OperationNode {
    match cell.language {
        SupportedLanguage::PyO3 => {
            // TODO: all callable functions should be exposed as their own OperationNodes
            //       these can be depended on throughout the graph as their own cells.
            let paths =
                chidori_static_analysis::language::python::parse::extract_dependencies_python(
                    &cell.source_code,
                );
            let report = chidori_static_analysis::language::python::parse::build_report(&paths);

            let mut input_signature = InputSignature::new();
            for (key, value) in &report.cell_depended_values {
                input_signature.globals.insert(
                    key.clone(),
                    InputItemConfiguation {
                        ty: Some(InputType::String),
                        default: None,
                    },
                );
            }

            let mut output_signature = OutputSignature::new();
            for (key, value) in &report.cell_exposed_values {
                output_signature.globals.insert(
                    key.clone(),
                    OutputItemConfiguation {
                        ty: Some(InputType::String),
                    },
                );
            }

            for (key, value) in &report.triggerable_functions {
                output_signature.functions.insert(
                    key.clone(),
                    OutputItemConfiguation {
                        ty: Some(InputType::Function),
                    },
                );
            }

            let cell = cell.clone();
            OperationNode::new(
                cell.name.clone(),
                input_signature,
                output_signature,
                Box::new(move |s, x, _, _| {
                    let cell = cell.clone();
                    // TODO: this needs to handle stdout and errors
                    let s = s.clone();
                    async move {
                        crate::library::std::code::runtime_pyo3::source_code_run_python(
                            &s,
                            &cell.source_code,
                            &x,
                            &cell.function_invocation,
                        ).await
                            .unwrap()
                            .0
                    }.boxed()
                }),
            )
        }
        SupportedLanguage::Starlark => OperationNode::new(
            cell.name.clone(),
            InputSignature::new(),
            OutputSignature::new(),
            Box::new(|_, x, _, _| async move { x}.boxed()),
        ),
        SupportedLanguage::Deno => {
            let paths =
                chidori_static_analysis::language::javascript::parse::extract_dependencies_js(
                    &cell.source_code,
                );
            let report = chidori_static_analysis::language::javascript::parse::build_report(&paths);

            let mut input_signature = InputSignature::new();
            for (key, value) in &report.cell_depended_values {
                input_signature.globals.insert(
                    key.clone(),
                    InputItemConfiguation {
                        ty: Some(InputType::String),
                        default: None,
                    },
                );
            }

            let mut output_signature = OutputSignature::new();
            for (key, value) in &report.cell_exposed_values {
                output_signature.globals.insert(
                    key.clone(),
                    OutputItemConfiguation {
                        ty: Some(InputType::String),
                    },
                );
            }

            for (key, value) in &report.triggerable_functions {
                output_signature.functions.insert(
                    key.clone(),
                    OutputItemConfiguation {
                        ty: Some(InputType::Function),
                    },
                );
            }

            let cell = cell.clone();
            OperationNode::new(
                cell.name.clone(),
                input_signature,
                output_signature,
                Box::new(move |_, x, _, _| {
                    // TODO: this needs to handle stdout and errors
                    let cell = cell.clone();
                    async move {
                        let result = tokio::task::spawn_blocking(move || {
                            let runtime = tokio::runtime::Runtime::new().unwrap();
                            let result = runtime.block_on(crate::library::std::code::runtime_deno::source_code_run_deno(
                                &ExecutionState::new(),
                                &cell.source_code,
                                &x,
                                &cell.function_invocation,
                            ));
                            match result {
                                Ok(v) => v.0,
                                Err(e) => panic!("{:?}", e),
                            }
                        }).await.unwrap();
                        result
                    }.boxed()
                }),
            )
        }
    }
}


#[cfg(test)]
mod test {
    #[tokio::test]
    async fn test_code_cell() {


    }
}