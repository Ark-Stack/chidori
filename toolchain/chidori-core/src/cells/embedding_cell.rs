use futures_util::FutureExt;
use crate::cells::{LLMEmbeddingCell, LLMPromptCell, TextRange};
use crate::execution::primitives::operation::{InputItemConfiguration, InputSignature, InputType, OperationFnOutput, OperationNode, OutputItemConfiguration, OutputSignature};
use crate::library::std::ai::llm::ai_llm_run_embedding_model;

#[tracing::instrument]
pub fn llm_embedding_cell(cell: &LLMEmbeddingCell, range: &TextRange) -> anyhow::Result<OperationNode> {
    let LLMEmbeddingCell {
        function_invocation,
        configuration,
        req,
        name,
        ..
    } = cell;
    let schema =
        chidori_prompt_format::templating::templates::analyze_referenced_partials(&&req);
    let mut role_blocks =
        chidori_prompt_format::templating::templates::extract_roles_from_template(&&req);

    let mut output_signature = OutputSignature::new();
    if let Some(name) = name {
        output_signature.functions.insert(
            name.clone(),
            OutputItemConfiguration::Value,
        );
    }

    let mut input_signature = InputSignature::new();

    // We only require the globals to be passed in if the user has not specified this prompt as a function
    if configuration.get("fn").is_none() {
        for (key, value) in &schema.items {
            input_signature.globals.insert(
                key.clone(),
                InputItemConfiguration {
                    ty: Some(InputType::String),
                    default: None,
                },
            );
        }
    }

    let (_, template) = role_blocks.drain(..).next().unwrap();
    let template = template.expect("Must include template");
    let name = name.clone();
    let is_function_invocation = function_invocation.clone();
    Ok(OperationNode::new(
        name.clone(),
        input_signature,
        output_signature,
        Box::new(move |s, x, _, _|  {
            let template = template.clone();
            let name = name.clone();
            let s = s.clone();
            async move {
                Ok(OperationFnOutput::with_value(ai_llm_run_embedding_model(&s, x, template, name.clone(), is_function_invocation).await))
            }.boxed()
        }),
    ))
}
