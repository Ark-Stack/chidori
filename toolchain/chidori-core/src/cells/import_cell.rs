use futures_util::FutureExt;
use tokio::runtime::Runtime;
use crate::cells::WebserviceCell;
use crate::execution::primitives::operation::{InputItemConfiguration, InputSignature, InputType, OperationNode, OutputSignature};
use crate::execution::primitives::serialized_value::{RkyvObjectBuilder, RkyvSerializedValue as RKV};


/// Import cells are provided a list of addresses to fetch dependencies from
/// when imported - we pull in a set of cells from a remote source - import cells are namespaced.

