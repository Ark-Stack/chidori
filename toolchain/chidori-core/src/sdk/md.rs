use crate::sdk::entry::Chidori;
use chidori_prompt_format::extract_yaml_frontmatter_string;
use indoc::indoc;
use std::collections::HashMap;
use std::path::Path;
use serde_derive::Serialize;
use crate::cells::{CellTypes, CodeCell, LLMCodeGenCell, LLMEmbeddingCell, LLMPromptCell, MemoryCell, SupportedLanguage, SupportedMemoryProviders, SupportedModelProviders, TemplateCell, WebserviceCell};

#[derive(PartialEq, Serialize, Debug)]
pub struct MarkdownCodeBlock {
    pub tag: String,
    pub name: Option<String>,
    pub body: String,
}

pub struct ParsedFile {
    // filename: Box<PathBuf>,
    // code: String,
    num_lines: usize,
    pub(crate) result: Vec<MarkdownCodeBlock>,
}

pub(crate) fn extract_code_blocks(body: &str) -> Vec<MarkdownCodeBlock> {
    let parts: Vec<&str> = body.split("```").collect();

    let mut code_blocks = Vec::new();
    for (index, part) in parts.iter().enumerate() {
        // Code blocks are at odd indices after splitting by ```
        if index % 2 == 1 {
            code_blocks.push(*part);
        }
    }

    code_blocks
        .iter()
        .map(|m| m.trim().to_string())
        .map(|m| {
            let mut lines = m.lines();
            let first_line = lines.next().unwrap_or_default();
            let rest: String = lines.collect::<Vec<&str>>().join("\n");

            // Extract the name in parentheses from the first line
            let tag_and_name: Vec<&str> = first_line.split_whitespace().collect();
            let tag = tag_and_name.get(0).cloned().unwrap_or_default().to_string();
            let name = tag_and_name.get(1).and_then(|n| n.strip_prefix('(').and_then(|n| n.strip_suffix(')').and_then(|n| Some(n.to_string()))));

            MarkdownCodeBlock {
                tag,
                name,
                body: rest,
            }
        })
        .collect()
}


fn parse_markdown_file(filename: &Path) -> ParsedFile {
    match std::fs::read_to_string(filename) {
        Err(e) => ParsedFile {
            // filename: Box::new(filename.to_path_buf()),
            // code: "".to_owned(),
            num_lines: 0,
            result: vec![],
        },
        Ok(source) => {
            let num_lines = source.lines().count();
            let result = extract_code_blocks(&source);
            ParsedFile {
                // filename: Box::new(filename.to_path_buf()),
                // code: source.to_string(),
                num_lines,
                result,
            }
        }
    }
}

pub fn load_folder(path: &Path) -> anyhow::Result<Vec<ParsedFile>> {
    let mut res = vec![];
    for entry in path.read_dir()? {
        let entry = entry?;
        let metadata = entry.metadata()?;

        let path = entry.path();
        if metadata.is_dir() {
            res.extend(load_folder(&path)?);
        }

        if metadata.is_file() && path.extension().and_then(|s| s.to_str()) == Some("md") {
            let parsed_file = parse_markdown_file(&path);
            res.push(parsed_file);
        }
    }
    Ok(res)
}

pub fn interpret_code_block(block: &MarkdownCodeBlock) -> Option<CellTypes> {
    let (frontmatter, body) = chidori_prompt_format::templating::templates::split_frontmatter(&block.body).unwrap();
    match block.tag.as_str() {
        "python" | "javascript" => {
            let language = match block.tag.as_str() {
                "python" => SupportedLanguage::PyO3,
                "javascript" => SupportedLanguage::Deno,
                _ => unreachable!(), // Given the outer match, this branch should never be reached
            };
            Some(CellTypes::Code(CodeCell {
                name: block.name.clone(),
                language,
                source_code: block.body.clone(),
                function_invocation: None,
            }))
        },
        "memory" => Some(CellTypes::Memory(MemoryCell {
            name: block.name.clone(),
            provider: SupportedMemoryProviders::InMemory,
            embedding_function: block.body.clone(),
        })),
        "embedding" => Some(CellTypes::Embedding(LLMEmbeddingCell {
            function_invocation: false,
            configuration: serde_yaml::from_str(&frontmatter).unwrap(),
            name: block.name.clone(),
            req: body,
        })),
        "prompt" => Some(CellTypes::Prompt(LLMPromptCell::Chat {
            function_invocation: false,
            configuration: serde_yaml::from_str(&frontmatter).unwrap(),
            name: block.name.clone(),
            provider: SupportedModelProviders::OpenAI,
            req: body,
        })),
        "codegen" => Some(CellTypes::CodeGen(LLMCodeGenCell {
            function_invocation: false,
            configuration: serde_yaml::from_str(&frontmatter).unwrap(),
            name: block.name.clone(),
            provider: SupportedModelProviders::OpenAI,
            req: body,
        })),
        "html" => Some(CellTypes::Template(TemplateCell {
            name: block.name.clone(),
            body: block.body.clone(),
        })),
        "web" => Some(CellTypes::Web(WebserviceCell {
            name: block.name.clone(),
            configuration: block.body.clone(),
            port: serde_yaml::from_str::<HashMap<String,String>>(&frontmatter).unwrap().get("port").and_then(|p| p.parse::<u16>().ok()).or_else(|| Some(8080)).unwrap(),
        })),
        _ => None,
    }
}


#[cfg(test)]
mod test {
    use super::*;
    use crate::execution::primitives::serialized_value::RkyvObjectBuilder;
    use indoc::indoc;
    use std::collections::HashMap;
    use std::fs;

    #[test]
    fn test_core1() {
        let contents = fs::read_to_string("./examples/core1_simple_math/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents).iter().map(interpret_code_block).collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_core2() {
        let contents = fs::read_to_string("./examples/core2_marshalling/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents).iter().map(interpret_code_block).collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_core3() {
        let contents = fs::read_to_string("./examples/core3_function_invocations/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents).iter().map(interpret_code_block).collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_core4() {
        let contents = fs::read_to_string("./examples/core4_async_function_invocations/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents).iter().map(interpret_code_block).collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_core5() {
        let contents = fs::read_to_string("./examples/core5_prompts_invoked_as_functions/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents).iter().map(interpret_code_block).collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_core6() {
        let contents = fs::read_to_string("./examples/core6_prompts_leveraging_function_calling/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents).iter().map(interpret_code_block).collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_core7() {
        let contents = fs::read_to_string("./examples/core7_rag_stateful_memory_cells/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents).iter().map(interpret_code_block).collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_extract_markdown() {
        let extracted = extract_code_blocks(indoc! {  r#"
        Generation
        
        ```python
        y = 20
        def add(a, b):
            return a + b
        ```

        ```javascript
        ---
        a: 2
        ---
        const x = add(2,2);
        ```
        
        ```prompt (multi_prompt)
        Multiply {y} times {x}
        ```

        ```html (named_html)
        <div>Example</div>
        ```
        "#
        });

        let mut map = HashMap::new();
        map.insert("a".to_string(), "2".to_string());
        assert_eq!(
            extracted,
            vec![
                MarkdownCodeBlock {
                    tag: "python".to_string(),
                    name: None,
                    body: indoc! { r#"
                y = 20
                def add(a, b):
                    return a + b"#}
                    .to_string(),
                },
                MarkdownCodeBlock {
                    tag: "javascript".to_string(),
                    name: None,
                    body: indoc! { r#"
                    ---
                    a: 2
                    ---
                    const x = add(2,2);"#}
                    .to_string(),
                },
                MarkdownCodeBlock {
                    tag: "prompt".to_string(),
                    name: Some("multi_prompt".to_string()),
                    body: "Multiply {y} times {x}".to_string(),
                },
                MarkdownCodeBlock {
                    tag: "html".to_string(),
                    name: Some("named_html".to_string()),
                    body: "<div>Example</div>".to_string(),
                }
            ]
        );
    }
}
