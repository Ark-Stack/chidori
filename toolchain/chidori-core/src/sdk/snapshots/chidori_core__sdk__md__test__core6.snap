---
source: chidori-core/src/sdk/md.rs
expression: v
---
- Code:
    name: math_fn
    language: PyO3
    source_code: "def add_two_numbers(a, b):\n      return a + b"
    function_invocation: ~
- Prompt:
    Chat:
      function_invocation: false
      configuration:
        import:
          - add_two_numbers
        fn: add_population
        model: gpt-3.5-turbo
        frequency_penalty: ~
        max_tokens: ~
        presence_penalty: ~
        stop: ~
        temperature: ~
        logit_bias: ~
        user: ~
        seed: ~
        top_p: ~
      name: add_population
      provider: OpenAI
      req: "Add the population of {{state}} to the population of California"
- Code:
    name: entry
    language: PyO3
    source_code: "import unittest\n\nclass TestMarshalledValues(unittest.IsolatedAsyncioTestCase):\n    async def test_run_prompt(self):\n        self.assertEqual(await add_population(state=\"Arizona\"), 4)\n\nunittest.TextTestRunner().run(unittest.TestLoader().loadTestsFromTestCase(TestMarshalledValues))"
    function_invocation: ~
